internal/batch/batch_processor_test.go

```diff
@@ -423,6 +423,62 @@ func TestAddWorkBatchOfOne(t *testing.T) {
 	}, bp.assemblyQueue)
 }
 
+func TestAddWorkDifferentKeys(t *testing.T) {
+	cancel, _, bp := newTestBatchProcessor(t, func(c context.Context, state *DispatchPayload) error {
+		return nil
+	})
+	defer cancel()
+
+	msg1 := &core.Message{
+		Sequence: 200,
+		Header: core.MessageHeader{
+			SignerRef: core.SignerRef{
+				Key: "0x1",
+			},
+		},
+	}
+	msg2 := &core.Message{
+		Sequence: 201,
+		Header: core.MessageHeader{
+			SignerRef: core.SignerRef{
+				Key: "0x2",
+			},
+		},
+	}
+	msg3 := &core.Message{
+		Sequence: 202,
+		Header: core.MessageHeader{
+			SignerRef: core.SignerRef{
+				Key: "0x1",
+			},
+		},
+	}
+
+	full, overflow := bp.addWork(&batchWork{msg: msg1})
+	assert.False(t, full)
+	assert.False(t, overflow)
+	assert.Equal(t, []*batchWork{
+		{msg: msg1},
+	}, bp.assemblyQueue)
+
+	full, overflow = bp.addWork(&batchWork{msg: msg3})
+	assert.False(t, full)
+	assert.False(t, overflow)
+	assert.Equal(t, []*batchWork{
+		{msg: msg1},
+		{msg: msg3},
+	}, bp.assemblyQueue)
+
+	full, overflow = bp.addWork(&batchWork{msg: msg2})
+	assert.True(t, full)
+	assert.True(t, overflow)
+	assert.Equal(t, []*batchWork{
+		{msg: msg1},
+		{msg: msg3},
+		{msg: msg2},
```

Comment:
I added this test to specifically verify and call out the ordering here. Because of the way the batch is broken up, the messages are not dispatched strictly in sequence order. I think this is OK but may need to double-check on assumptions about sequence on both sender and receiver.

Reply:
I _think_ the same scenario can happen even if all 3 messages were signed with the same key, but somehow `msg2` "misses the train" due to random timing and the batch goes out with just `msg1` and `msg3`. This means the order in which nonces are allocated will not match the order of the local message `sequence`. Just trying to assure myself that this is alright.

Reply:
There are other existing tests in this file (a few lines before this one) which simulate messages being added to a batch out of sequence order. I assumed this was because we dispatch messages randomly to a pool of message writers, and there can be a delay between when a message grabs a sequence number and when it's actually committed in a database transaction. Otherwise I'm not sure of the reason for the existing logic (and unit tests) around sorting messages as they get assembled into a batch.

Reply:
So here's the logic... and yes 👍 to exactly your explanation for the sequence ordering. Thanks for the detail, and you're right I did need to go back and look....

![image](https://github.com/hyperledger/firefly/assets/6660217/e3c059d2-57f0-47ca-bc43-45999e841b95)

One note, is that the ordering only clicks in when you're writing messages _in parallel_. So on the FireFly core API you're not waiting for message 1 to be written, before message 2. So the fact that ordering in this scenario is not deterministic was accepted, because it can't be deterministic if it's parallel.

That's my recollection looking back at the diagram. Thanks for